### DISCUSSION TITLE
Handling electrons with Kinetic Energy above 1 MeV

### DISCUSSION THREAD

**Discussion by sdevkaushik** [topas-nbio/TOPAS-nBio#14] [Simulation-question]
### Simulation question

Hi, I am transforming a Geant4 code in TOPAS. The code simulates an ICRU sphere with a scorer at the center of the sphere (actually there are three scorers but let’s assume one scorer at the center of the sphere for simplicity). The ICRU sphere is irradiated with 10 MeV monoenergetic neutrons using a surface source placed at the surface of the sphere. The secondary species generated are recorded by the scorer. In TOPAS, I am using TsNtupleForCell scorer. The spectra of secondary species will be fed into TOPAS-nBio. In the Geant4 version, they have handled electrons with energy above 1 MeV and below 1 MeV separately. The electrons generated with energies below 1 MeV in the scorer were killed and recorded. The electrons above 1 MeV were allowed to track down to 1 MeV and if they happened to interact in the scorer their energy deposition was recorded with Geant4 and then killed and passed to Geant4-DNA as independent tracks. I have modified the TsNtupleForCell to implement the same. 

Here are the Geant4 classes (SpectraSD.cc and SteppingAction.cc) that handle electron interaction. 

SpectraSD.cc
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

G4bool SpectraSD::ProcessHits(G4Step* step, G4TouchableHistory*)
{
    // G4double edep = step->GetTotalEnergyDeposit();
    // if ( edep == 0. ) return true;

    // G4cout << "A step happened." << G4endl;

    // Get the number of secondaries produced in the current step
    G4int nSec = step->GetNumberOfSecondariesInCurrentStep();
    if ( nSec == 0. ) return true;

    // G4cout << nSec << " secondaries produced in this region before processing" << G4endl;

    // Get secondaries produced in the current step - not the same as GetSecondary
    // This method starts with GetSecondary, then returns the sub-list that makes up its last n entries,
    //      where n is the number of secondaries produced in that step. It's weird.
    const std::vector<const G4Track*>* secondaries = step->GetSecondaryInCurrentStep();

    // Start an instance of the GenericIonsManager to get the recoil nuclei
    

    // Record name and kinetic energy of each of the secondary particles of interest - but only record
    //      if those particles are present
    for (G4int i=0; i<nSec; i++) {
        const G4Track* track = (*secondaries)[i];
        const G4ParticleDefinition* particle = track->GetParticleDefinition();
        G4String pName = particle->GetParticleName();
        G4double eKin = track->GetKineticEnergy();

        // >1 MeV electrons are handled separately
        if ( pName != "e-" || eKin <= 1.*MeV ) {

            SpectraHit* hit = new SpectraHit(pName, eKin);
            fHitsCollection->insert(hit);

        }

        // G4cout << pName << " produced with energy " << G4BestUnit(eKin, "Energy") << " in SpectraSD" << G4endl;

    }

    
    
    return true;
}

SteppingAction.cc

void SteppingAction::UserSteppingAction(const G4Step* aStep)
{
    // Kill electrons below 1 MeV and, if they deposit energy or are killed within a scoring volume,
    // add their energy to the appropriate histogram
    const G4ParticleDefinition* particleType = aStep->GetTrack()->GetParticleDefinition();
    G4String pName = particleType->GetParticleName();
    if ( pName == "e-" ) {
    
        //G4cout << pName << " pre: " << aStep->GetPreStepPoint()->GetKineticEnergy()/keV << " keV" << G4endl;
        
        // Interested in the energy at the end of the step
        G4double eKin = aStep->GetPostStepPoint()->GetKineticEnergy();
        //G4cout << pName << " post: " << eKin/keV << " keV" << G4endl;
        if ( eKin <= 1.*MeV ) {

            aStep->GetTrack()->SetTrackStatus(fStopAndKill);

           //G4cout << "Killed and TBR if in scoring volume" << G4endl;
           
            // Use the PreStepPoint to get volume information
            G4String volName = aStep->GetPreStepPoint()->GetPhysicalVolume()->GetName();
            if ( volName == "Target1" ) {
                fRunAction->Sume1Dose(eKin/eV);
                fRunAction->Sum1Dose(eKin/eV);
                G4AnalysisManager::Instance()->FillH1(9,eKin/eV);
                
                //G4cout << "An electron with energy " << eKin/keV << " keV was killed in Target1" << G4endl;
            }
            else if ( volName == "Target2" ) {
                fRunAction->Sume2Dose(eKin/eV);
                fRunAction->Sum2Dose(eKin/eV);
                G4AnalysisManager::Instance()->FillH1(21,eKin/eV);
                
                //G4cout << "An electron with energy " << eKin/keV << " keV was killed in Target2" << G4endl;
            }
            else if ( volName == "Target3" ) {
                fRunAction->Sume3Dose(eKin/eV);
                fRunAction->Sum3Dose(eKin/eV);
                G4AnalysisManager::Instance()->FillH1(33,eKin/eV);
                
                //G4cout << "An electron with energy " << eKin/keV << " keV was killed in Target3" << G4endl;
            }

        }

    }

Here is TsNtupleForCell.cc that I have modified to implement the same in TOPAS.
// Scorer for TsNtupleForCellNew
//
// ********************************************************************
// *                                                                  *
// * This file is part of the TOPAS-nBio extensions to the            *
// *   TOPAS Simulation Toolkit.                                      *
// * The TOPAS-nBio extensions are freely available under the license *
// *   agreement set forth at: https://topas-nbio.readthedocs.io/     *
// *                                                                  *
// ********************************************************************
//
//Tuple scorer for scoring events deposited in the cell organelles

#include "TsNtupleForCellNew.hh"

#include "G4SystemOfUnits.hh"

TsNtupleForCellNew::TsNtupleForCellNew(TsParameterManager* pM, TsMaterialManager* mM, TsGeometryManager* gM, TsScoringManager* scM, TsExtensionManager* eM,
                                 G4String scorerName, G4String quantity, G4String outFileName, G4bool isSubScorer)
: TsVNtupleScorer(pM, mM, gM, scM, eM, scorerName, quantity, outFileName, isSubScorer)
{
    //SetScorer();
    fNtuple->RegisterColumnF(&fPosX, "Position X", "cm");
    fNtuple->RegisterColumnF(&fPosY, "Position Y", "cm");
    fNtuple->RegisterColumnF(&fPosZ, "Position Z", "cm");
    fNtuple->RegisterColumnF(&fEnergy, "Energy", "MeV");
    fNtuple->RegisterColumnF(&fEnergyPostStep, "PostStepKineticEnergy", "MeV");
    fNtuple->RegisterColumnF(&fEnergyDep, "Energy Deposited", "MeV");
    fNtuple->RegisterColumnI(&fParticleType, "Particle Type (in PDG Format)");
    fNtuple->RegisterColumnI(&fTrackID, "Track ID");
    fNtuple->RegisterColumnI(&fRunID, "Run ID");
    fNtuple->RegisterColumnI(&fEventID, "Event ID");
    fNtuple->RegisterColumnS(&fVolName, "Volume Name");

}

TsNtupleForCellNew::~TsNtupleForCellNew() {;}

G4bool TsNtupleForCellNew::ProcessHits(G4Step* aStep, G4TouchableHistory*)
{
    ResolveSolid(aStep);
    
    G4double flagEnergyDep      = aStep->GetTotalEnergyDeposit();
    
    G4StepPoint* theStepPoint=aStep->GetPreStepPoint();

    G4StepPoint* theStepPointPost=aStep->GetPostStepPoint();
    
    //Find volume name
    G4Track* aTrack = aStep->GetTrack();
    G4String volumeName = aTrack->GetVolume()->GetName();
    
    //Get position
    G4ThreeVector pos = theStepPoint->GetPosition();

    //Get particle type
    fParticleType = aStep->GetTrack()->GetDefinition()->GetPDGEncoding();

    
    //Score events that deposit energy in the cell and it's organelles:
    if ((flagEnergyDep > 0 && fParticleType != 11)){
        
        //Get position
        fPosX = pos.x();
        fPosY = pos.y();
        fPosZ = pos.z();
        
        //Get particle Energy
        fEnergy = theStepPoint->GetKineticEnergy();
	
	//Get post step Kinetic Energy
	fEnergyPostStep = theStepPointPost->GetKineticEnergy();
     
        //Get Edep
        fEnergyDep = flagEnergyDep;
       
	//Get particle type
	fParticleType = aStep->GetTrack()->GetDefinition()->GetPDGEncoding();

        // Get IDs
        fTrackID = aStep->GetTrack()->GetTrackID();
        fRunID   = GetRunID();
        fEventID = GetEventID();
        
        //Get volume Name
        fVolName = volumeName;

        fNtuple->Fill();

        return true;
    }

    if (( fParticleType == 11 && fEnergyPostStep <= 1.*MeV)){
        
        //Get position
        fPosX = pos.x();
        fPosY = pos.y();
        fPosZ = pos.z();
        
        //Get particle Energy
        fEnergy = theStepPoint->GetKineticEnergy();

	
	//Get post step Kinetic Energy
	fEnergyPostStep = theStepPointPost->GetKineticEnergy();
     

        //Get Edep
        fEnergyDep = flagEnergyDep;
       
	//Get particle type
	fParticleType = aStep->GetTrack()->GetDefinition()->GetPDGEncoding();

        // Get IDs
        fTrackID = aStep->GetTrack()->GetTrackID();
        fRunID   = GetRunID();
        fEventID = GetEventID();
        
        //Get volume Name
        fVolName = volumeName;

	aStep->GetTrack()->SetTrackStatus(fStopAndKill);

        fNtuple->Fill();

        return true;
    }
    
 
    return false;   
}

I would appreciate any suggestions.

### TOPAS-nBio category

Scoring

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Attestation

- [X] I testify that I have fully read the bug reporting guideline beforehand.

**Comment by JoseRamosMendez** [comment]
Could you be more explicit about what topic suggestions you need?

**Reply by sdevkaushik** [reply]
Thanks for replying. I need suggestions about the scoring of electrons with energy greater than 1 MeV in the TsNtupleForCell scorer.
Here are a few cases that can arise while tracking electrons.

(i) Any electron having energy less than 1 MeV generated in the scorer or generated elsewhere in the geometry passing the scorer should be killed and recorded.
(ii) Any electron of energy greater than 1 MeV generated in the scorer should be tracked down and its energy deposition in the scorer should be recorded until it leaves the scorer. If before leaving the scorer, the electron energy becomes less than 1 MeV then kill it and record it. 
(iii) Any electron of energy greater than 1 MeV generated outside the scorer, crossing the scorer and leaving with energy greater than 1 MeV. In this case, record the energy deposition in the scorer. 

To implement this, the Geant4 developer used a SteppingAction class. How can I do that in TOPAS with TsNtupleForCell scorer or is there any other way to implement this?

**Comment by sdevkaushik** [comment]
![ElectronsComparisonTsNtupleForCellNew](https://github.com/topas-nbio/TOPAS-nBio/assets/167898674/a9a2104f-c0fa-4fb1-ae91-e7360837acaa)
Hi, I am using TsNtupleForCell scorer. I have modified it to stop electrons below 1 MeV. I have already shared my scorer in this thread. However, TsNtupleForCell scorer is not stopping electrons below 1 MeV instead it is stopping electron below 2 MeV. For reference, please see the attached graph. Is it a default TsNtupleForCell scorer parameter? How can I stop electrons below 1 MeV?

**Comment by JoseRamosMendez** [comment]
Hi there,
I guess you're excluding neutrons (they also lose energy via hadronic interactions). So, considering only e-, I think you can put everything within the same if conditional. Also use the kinetic energy at the **preStepPoint**, not postStepPoint, e.g.,

    if ( flagEnergyDep > 0 && (fParticleType == 11 && fEnergyPreStep <= 1.*MeV )) {
            ....
            ....
            aStep->GetTrack()->SetTrackStatus(fStopAndKill);
            fTuple->Fill();
            return true;
        }
    }

The `ProcessHit` works in a different way than the `SteppingAction`. `ProcessHits` works only in the assigned Component whereas `SteppingAction` works everywhere. Thus, as long you assign the scorer to the right component, the `ProcessHits` "won't see" what happens outside of it. So, you have to create another scorer to just kill electrons < 1 MeV, and attach it to the outer components of your scorer.

**Reply by sdevkaushik** [reply]
Hi, 
thanks for replying, and thanks for the explanations. 
I was able to match spectra with Geant4 by modifying the scorer I shared previously.

I used the following conditions in my new scorer:
 
if (flagEnergyDep > 0 || (fParticleType != 11 && fEnergyPostStep <=1.*MeV)){
        
        //Get position
        fPosX = pos.x();
        fPosY = pos.y();
        fPosZ = pos.z();
        
        //Get particle Energy
        fEnergy = theStepPoint->GetKineticEnergy();
	
	//Get post step Kinetic Energy
	fEnergyPostStep = theStepPointPost->GetKineticEnergy();
     
        //Get Edep
        fEnergyDep = flagEnergyDep;
       
	//Get particle type
	fParticleType = aStep->GetTrack()->GetDefinition()->GetPDGEncoding();

        // Get IDs
        fTrackID = aStep->GetTrack()->GetTrackID();
        fRunID   = GetRunID();
        fEventID = GetEventID();
        
        //Get volume Name
        fVolName = volumeName;
	
	//G4cout<< fParticleType << " produced with energy "<< fEnergy << " in " << fVolName<<G4endl;

       
   
	if(( fParticleType == 11 && fEnergyPostStep <= 1.*MeV)){
	  
	  aStep->GetTrack()->SetTrackStatus(fStopAndKill);
	  
	  
	 // G4cout<<" An electron with energy "<< fEnergy << " was killed in "<<fVolName<<G4endl;
	 
	} 
	  fNtuple->Fill();
	  
	  
	  return true;
    }

However, now the electrons are not being stopped at 1 MeV instead they are being stopped at 2 MeV. Please see the attached plot and comparison with Geant4 in my last post. Why is this happening? 

I am interested in electron energy at the end of the step so I have used postStepPoint condition.

I don't want to exclude neutrons and their contribution to energy deposition. I have used the following condition in my code:
s:Sc/InnerScorer/OnlyIncludeParticlesOfGeneration = "secondary"
I can remove this statement also to include neutrons.

**Comment by sdevkaushik** [comment]
I was able to stop electrons at 1 MeV using TsNtupleForCell scorer.
